backends/p4tools/testgen/core/small_step/expr_stepper.cpp

```diff
@@ -0,0 +1,475 @@
+    if (selectExpression->selectCases.empty()) {
+        stepNoMatch();
+        return false;
+    }
+
+    const auto* selectCase = selectExpression->selectCases.at(0);
+    // Getting P4ValueSet from PathExpression , to highlight a particular case of processing
+    // P4ValueSet.
+    if (const auto* pathExpr = selectCase->keyset->to<IR::PathExpression>()) {
+        const auto* decl = state.findDecl(pathExpr)->getNode();
+        // Until we have test support for value set, we simply remove them from the match list.
+        // TODO: Implement value sets.
+        if (decl->is<IR::P4ValueSet>()) {
+            auto* newSelectExpression = selectExpression->clone();
+            newSelectExpression->selectCases.erase(newSelectExpression->selectCases.begin());
+            auto* nextState = new ExecutionState(state);
+            nextState->replaceTopBody(Continuation::Return(newSelectExpression));
+            result->emplace_back(nextState);
+            return false;
+        }
+    }
+
+    if (!SymbolicEnv::isSymbolicValue(selectCase->keyset)) {
+        // Evaluate the keyset in the first select case.
+        return stepToSubexpr(selectCase->keyset, result, state,
+                             [selectExpression, selectCase](const Continuation::Parameter* v) {
+                                 auto* newSelectCase = selectCase->clone();
+                                 newSelectCase->keyset = v->param;
+
+                                 auto* result = selectExpression->clone();
+                                 result->selectCases[0] = newSelectCase;
+                                 return Continuation::Return(result);
+                             });
+    }
+
+    if (!SymbolicEnv::isSymbolicValue(selectExpression->select)) {
+        // Evaluate the expression being selected.
+        return stepToListSubexpr(selectExpression->select, result, state,
+                                 [selectExpression](const IR::ListExpression* listExpr) {
+                                     auto* result = selectExpression->clone();
+                                     result->select = listExpr;
+                                     return Continuation::Return(result);
+                                 });
+    }
+
+    // TODO: Implement the taint case for select expressions.
+    // In the worst case, this means the entire parser is tainted.
+    if (state.hasTaint(selectExpression->select) || state.hasTaint(selectCase->keyset)) {
+        TESTGEN_UNIMPLEMENTED(
+            "The SelectExpression %1% is trying to match on a tainted key set."
+            " This means it is matching on uninitialized data."
+            " P4Testgen currently does not support this case.",
+            selectExpression);
+    }
+
+    // Handle case where the first select case matches: proceed to the next parser state,
+    // guarded by its path condition.
+    const auto* equality = GenEq::equate(selectExpression->select, selectCase->keyset);
```

Comment:
keyset can represent a set, not just a value. I hope that equate handles this.
Otherwise you need something like "ismember"

Reply:
There is a PR in-flight that rewrites this function, which should also include simplifications. There is some more issues with it. For example, default expressions and taint expressions are not handled quite correctly.